content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,339 @@ Like a segment tree on values rather than indices.
 	</Resource>
 </Resources>
 
+## Introduction
+Suppose you want to support the following queries:
+
+- Given a range, count the number of occurrences of value $x$.
+- Given a range, find the $k$ smallest element
+
+With a wavelet tree, you can easily support these queries in $\mathcal{O}(\log M)$ time,
+where $M$ is the maximum value in the array.
+
+### Wavelet Tree Structure
+
+To answer value-based queries efficiently, we'll create a segment tree where each node represents a range of values, instead of indices. Just like a normal segment tree, each subsequent level splits the range into two halves. Note that an index can appear in at most $\log(M)$ nodes.
+
+<Spoiler title="Wavelet Tree Visualization">
+
+Let's say our array is: $[3,5,3,1,2,2,3,4,5,5]$
+Each node has an array representing the indices of every number between $l$ and $r$
+
+![Wavelet Tree Visualization](./assets/diagram.png)
+</Spoiler>
+
 ### Solution - Range K-th Smallest
+Before we solve this problem, let's consider a simpler version where we are asked, given a range, to count the number of occurrences of value $x$.
+
+## Solving the first type of query
+**Given a range $l$, $r$, count the number of occurrences of value x.**
+
+To calculate the number of occurrences from $𝑙$ to $𝑟$, we can use the following
+formula:
+
+$$
+\begin{aligned}
+\texttt{occurrences}(l, r) = \texttt{occurrences}(r) - \texttt{occurrences}(l)
+\end{aligned}
+$$
+
+This reduces the problem to counting the number of occurrences in a prefix.
+
+One way to solve the problem is to go to the leaf node
+and perform a binary search for the number of indices less than $𝑟$
+However, let's explore a different approach that can also be extended to the
+second type of query.
+
+Instead of binary searching on the leaf, we update $𝑟$ as we recurse down the
+tree.
+If we can determine the position (index) of $r$ in the left and right children of a node, we can recurse down the tree and determine its position in the leaf node.
+
+To find the position of $𝑟$ in a node's left and right children, we need to
+determine how many indices are smaller than the middle value (mid) and precede
+$𝑟$.
+This can be done using a prefix sum.
+
+Let's define:
+- $c[i]$ = as $1$ if $\texttt{index}$[$i$] is smaller than mid otherwise $0$
+- $\texttt{prefix\_b}$[$i$] as prefix sum of $c[i]$
+
+Formally
+
+$$
+c[i] = \begin{cases}
+    1, & \texttt{if } \texttt{index}[i] < \texttt{mid} \\
+    0, & \texttt{otherwise}
+\end{cases}
+$$
+$$
+\texttt{prefix\_b}[i] = \texttt{prefix\_b}[i - 1] + c[i]
+$$
+
+
+To update $r$ as we recurse down, we do the following:
+- To know the value of $r$ if we recurse left, we use $\texttt{prefix\_b}$[$r$]
+- If we recurse right, we use $r$ - $\texttt{prefix\_b}$[$r$]
+
+Now let's try to solve our main problem.
+## Solving the second type of query
+**Given a range $l$, $r$ find the k-th smallest element**
```

Comment:
```suggestion
```
also remove from "Solving the first type"

Reply:
"Where would they find information about the type of query we're trying to solve? Without the introduction, it would be quite difficult to locate. Wouldn't it be simpler to include a subtitle for that? Additionally, the first type of query is essential for solving the second type, but it isn't the focus of the problem, making it even harder to understand what we are trying to do.

Reply:
then make it part of the header itself

Reply:
In brackets? or as the header?